### PR TITLE
Delete deprecated backends

### PR DESCRIPTION
**Pull request by QuLogic**
## PR Summary

Remove GDK, GTK2, and WX backends, plus all deprecated methods in the backends. I need to go through a few of the examples to make sure they're okay.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by tacaswell** [comment]
Can we hold off on doing WX?  @DietmarSchwertberger is interested in keeping it up.

**Comment by QuLogic** [comment]
WX or WXAgg (which was not deprecated before)?

**Comment by QuLogic** [comment]
OK, I restored `backend_wx.py` for now as it's imported in `backend_wxagg.py`, so we still need it at least partially. I'll leave it to @DietmarSchwertberger to determine whether to remove anything from there.

There's one example `lineprops_dialog_gtk_sgskip.py` which uses `DialogLineprops` from the GTK2 backend, but this is not implemented in GTK3. I'm not sure whether this dialog should be implemented for GTK3 as well or the example deleted?

**Comment by DietmarSchwertberger** [comment]
Thanks. I would be fine with a deprecation message for 3.0 for backend_wx..
By then, things should be decided.

**Comment by QuLogic** [comment]
The WX renderer was already [deprecated in 2.0](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/backends/backend_wx.py#L190), but not WXAgg. However, `backend_wxagg.py` uses `backend_wx.py`. It doesn't look like the renderer is imported, so maybe only the renderer needs to be removed?

**Comment by DietmarSchwertberger** [comment]
Removing just the WX renderer would make `backend_wx.py` unusable on it's own, even though it could be imported.
Before 3.0 I would like to try whether the WX renderer can be used for rendering to wx.MetaFileDCs for creating EMF/WMF files. If that's not possible, it should be removed then. No matter what the outcome is, the WX backends will need restructuring for 3.0.
I would suggest not to do it for 2.2 as the risk for bugs is quite significant due to the tight schedule. (There are currently unmerged bugfix PRs.)

**Comment by QuLogic** [comment]
This PR _is_ about 3.0, but the wx changes are reverted in any case.

**Comment by DietmarSchwertberger** [comment]
Sorry, did not see this. I just had in mind that the Wx backend was deprecated for rev. 2 already.
I will take care for restructuring / clean up in time for 3.0.

File: examples/user_interfaces/pylab_with_gtk_sgskip.py [hunk 1]
```diff
@@ -23,32 +23,32 @@
 toolbar = manager.toolbar
 
 # now let's add a button to the toolbar
-import gtk
+import gi
+gi.require_version('Gtk', '3.0')
+from gi.repository import Gtk
```
Comment (anntzer) [inline examples/user_interfaces/pylab_with_gtk_sgskip.py]: rename "next" below to something else e.g. pos?

File: examples/widgets/cursor.py [hunk 2]
```diff
@@ -20,7 +20,7 @@
 ax.set_xlim(-2, 2)
 ax.set_ylim(-2, 2)
 
-# set useblit = True on gtkagg for enhanced performance
+# Set useblit=True on some backends for enhanced performance.
```
Comment (anntzer) [inline examples/widgets/cursor.py]: `"useblit=True" improves the performance on most backends`? (gtk3cairo (fundamentally) cannot support blitting and I'm not sure what the situation is with nbagg but all other major backends do support blitting.)

File: examples/widgets/span_selector.py [hunk 3]
```diff
@@ -38,7 +38,7 @@ def onselect(xmin, xmax):
     ax2.set_ylim(thisy.min(), thisy.max())
     fig.canvas.draw()
 
-# set useblit True on gtkagg for enhanced performance
+# Set useblit=True on some backends for enhanced performance.
```
Comment (anntzer) [inline examples/widgets/span_selector.py]: see above

File: lib/matplotlib/pyplot.py [hunk 4]
```diff
@@ -94,16 +94,11 @@ def _backend_selection():
...
-          backend not in ('GTK', 'GTKAgg', 'GTKCairo')):
-        if 'gi' in sys.modules:
-            from gi.repository import GObject
-            ml = GObject.MainLoop
-        else:
-            import gobject
-            ml = gobject.MainLoop
+    elif 'gtk' in sys.modules and 'gi' in sys.modules:
+        from gi.repository import GObject
+        ml = GObject.MainLoop
```
Comment (anntzer) [inline lib/matplotlib/pyplot.py]: squish into the next line?

**Comment by QuLogic** [comment]
Fixed comments.

**Comment by QuLogic** [comment]
Rebased due to conflicts.

**Comment by QuLogic** [comment]
Rebased due to conflicts.

**Comment by efiring** [comment]
It looks like we want to resolve the conflict by going with the master version, correct?

**Comment by QuLogic** [comment]
Partly, we can still drop this `Toolbar` class that was deprecated (and now removed) in all backends.

**Comment by QuLogic** [comment]
Rebased due to conflicts.

**Pull request merged**